consensus/src/transaction_shuffler/fairness/mod.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction_shuffler::{
+    fairness::{
+        conflict_key::ConflictKeyRegistry, conflict_zone::ConflictZone, pending_zone::PendingZone,
+    },
+    TransactionShuffler,
+};
+use aptos_types::transaction::SignedTransaction;
+use itertools::zip_eq;
+use selection_tracker::SelectionTracker;
+use std::collections::BTreeSet;
+
+pub(crate) mod conflict_key;
+mod conflict_zone;
+mod pending_zone;
+mod selection_tracker;
+
+#[cfg(test)]
+mod tests;
+
+type TxnIdx = usize;
+
+const NUM_CONFLICT_ZONES: usize = 3;
+
+#[derive(Debug)]
+struct FairnessShuffler {
+    sender_conflict_window_size: usize,
+    module_conflict_window_size: usize,
+    entry_fun_conflict_window_size: usize,
+}
+
+impl TransactionShuffler for FairnessShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let conflict_key_registries = ConflictKeyRegistry::build_registries(&txns);
+        let order = FairnessShufflerImpl::new(self, &conflict_key_registries).shuffle();
+        reorder(txns, &order)
+    }
+}
+
+fn reorder<T: Clone>(txns: Vec<T>, order: &[TxnIdx]) -> Vec<T> {
+    assert_eq!(txns.len(), order.len());
+    order.iter().map(|idx| txns[*idx].clone()).collect()
```

Comment:
Can we avoid the clone here? Cloning the txns is pretty expensive.

Reply:
Think about it, at least O(n) copying / cloning is needed and it's exactly what's done here.